feat(form-v2): add myInfo types and constants

seaerchin opened the pull request:
Part 1 of #3926

## Problem
Forms in react currently does not have any myinfo fields implemented, despite the angular version of the application supporting myinfo. This leads to a feature gap between the two versions of the application

Related to #3360 

## Solution
This PR adds MyInfo related constants and typings. These are as follows: 
1. drag/drop related constants
2. form preview related constants (the rendered form in the right panel) - this includes the label/icon etc 
3. adding a `MyInfoImplemented/MyInfoUnimplemented` type so that we encode type-safety into the methods and no callers can mistakenly call a myinfo method wrongly (this is because the figma has more myinfo fields than currently exists)

seaerchin committed:
chore(fields/base): add more myinfo attributes

chore(constants): add my info constants

seaerchin committed:
refactor(types/field): extend fields to have isMyInfo property

seaerchin committed:
feat(myinfo): add detailed myinfo types

revert(namefield): delete extra namefield

seaerchin committed:
refactor(types/field): rename myinfo symbol

seaerchin committed:
chore(admin-form): add myinfo constants

seaerchin committed:
refactor(myinfo): update myinfo typings

seaerchin committed:
feat(types/field): add DU for form fields

seaerchin committed:
refactor(types/field): update field types

seaerchin committed:
refactor(types): rejig types to fit myinfo fields

seaerchin committed:
refactor(myinfo): update myinfo related types and methods

seaerchin committed:
feat(field/base): refined myinfo typings to have implemented/unimplemented

seaerchin committed:
feat(constants): add remaining constants for myinfo

seaerchin committed:
fix(shared/types/field): fixed rebase retaining old code

seaerchin commented on shared/types/field/myinfo/base.ts:
```diff
@@ -0,0 +1,137 @@
+export const getMyInfoDropdownFieldMeta = (
+  myInfoAttribute: MyInfoDropdownTypes,
+): MyInfoFieldMeta<DropdownFieldBase> => {
+  const myInfoDropdownFields = _(MYINFO_FIELD_CONSTANTS)
```
i felt that this way (chaining) was cleaner rather than isolating the imports and writing it as 

```ts
const filteredFieldOptions = ... 
const myInfoDropdownFields = ...
return ...
```

but am open to changing if bundle size is an issue

karrui commented on frontend/src/features/admin-form/create/builder-and-design/constants.ts:
```diff
@@ -25,11 +29,69 @@ export const ALL_FIELDS_ORDERED = [
+export const CREATE_MYINFO_MARRIAGE_FIELDS_ORDERED =
+  MYINFO_FIELDS_ORDERED.slice(19, 24)
+
+export const CREATE_MYINFO_CHILDREN_FIELDS_ORDERED =
+  MYINFO_FIELDS_ORDERED.slice(32, 41)
+
+export const CREATE_MYINFO_INCOME_FIELDS_ORDERED =
+  MYINFO_FIELDS_ORDERED.slice(41)
```
these don't exist yet, should remove

seaerchin replied:
yeap, will do!

karrui commented on shared/types/field/myinfo/base.ts:
```diff
@@ -0,0 +1,137 @@
+import {
```
does this need to be in shared? if it's only used by the client we can just use this file there

seaerchin replied:
sure, will do! will relocate to `admin-form/create/builder-and-design`!

seaerchin replied:
relocated them to ^ and `~features/myinfo/`

karrui reviewed (approved):
lgtm, but please run some backend tests locally (they don't run on form-v2 branches), but there's quite a number of type changes so we should double check. may also want to backport this to develop branch.

karrui reviewed (changes requested):
haha i see types fail already, please fix them!~

seaerchin commented:
> haha i see types fail already, please fix them!~

oddly enough, i'm not seeing the file on this branch, which was why i thought it wasn't an issue; i think chromatic might have used the latest `form-v2/develop` and selected an outdated type. 

should i rebase and fix this?

seaerchin committed:
chore(myinfo/constants): remove unused constants

karrui commented:
> > haha i see types fail already, please fix them!~
> 
> 
> 
> oddly enough, i'm not seeing the file on this branch, which was why i thought it wasn't an issue; i think chromatic might have used the latest `form-v2/develop` and selected an outdated type. 
> 
> 
> 
> should i rebase and fix this?

Sure thing, rebase away

seaerchin committed:
chore(myinfo): deletes shared myinfo types and shift them to FE only

seaerchin committed:
chore(myinfo): add myinfo constants/types to FE

seaerchin committed:
chore(base): deletes extra myinfo types

seaerchin committed:
refactor(myinfo): update imports

seaerchin committed:
revert(myinfo): undo previous myinfo type changes to shared types

seaerchin committed:
fix(constants/myinfo): fixed typo

seaerchin commented:
rebased and fixed the chromatic error but GHA don't seem to be working (see #3940); i also reverted the `shared/type` changes so it's 1 less thing to worry about

karrui reviewed (approved):
lgtm. also reran the failing GA for you

seaerchin commented on shared/constants/field/myinfo/index.ts:
```diff
@@ -48,7 +48,7 @@ export const types: MyInfoFieldBlock[] = [
     source: 'Immigration & Checkpoints Authority / Ministry of Manpower',
     description:
-      'The registered name of the form-filler. This field is verified by ICA for Singaporeans/PRs & foreigners on Long-Term Visit Pass, and by MOM for Employment Pass holders.',
+      'The registered date of birth of the form-filler. This field is verified by ICA for Singaporeans/PRs & foreigners on Long-Term Visit Pass, and by MOM for Employment Pass holders.',
```
i found this while crawling through the codebase and it was last updated 22 months ago so i think it was always in the angularjs codebase @__@

seaerchin commented on shared/types/field/base.ts:
```diff
@@ -38,21 +41,41 @@ export enum MyInfoAttribute {
+  WorkpassType = 'workpasstype',
+  WorkpassStatus = 'workpassstatus',
   WorkpassExpiryDate = 'workpassexpirydate',
+  EmploymentSector = 'empsector',
+  HighestEducation = 'highested',
+  SchoolName = 'schname',
+  YearOfGraduation = 'gradyr',
+  ChildrenBirthCertificateNumber = 'childrenbirthcertno',
+  ChildrenName = 'childrenname',
+  ChildrenGender = 'childrengender',
+  ChildrenDateOfBirth = 'childrendob',
+  ChildrenLifeStatus = 'childrenlifestatus',
+  ChildrenRace = 'childrenrace',
+  ChildrenCountryOfBirth = 'childrencountryofbirth',
+  ChildrenResidentialStatus = 'childrenresidentialstatus',
+  ChildrenNationality = 'childrennationality',
+  CPFAccountBalance = 'cpfaccbal',
+  CPFContributionHistory = 'cpcontribhist',
+  NoticeOfAssessment = 'noticeofassessment',
+  OwnershipOfPrivateResidentialProperty = 'ownership',
```
shure, that works! should've clarified before embarking on a typing spree anyways HAHAHA

seaerchin committed:
chore(myinfoattribute): deletes unimplemented types

seaerchin committed:
chore(myinfo): update usages of impl/unimpl to base myinfoattr

karrui commented:
@seaerchin merge whenever

seaerchin committed:
feat(form-v2): myinfo implementation  (#3926)

* feat(fieldlistdrawer): add simple myinfo drawer

* feat(editfielddrawer): add placeholder for myinfo

* feat(builderanddesign): add section for myinfo personal section

* feat(fieldcreation): add new method for myinfo

* feat(editmyinfo): add editmyinfo component

* feat(fieldcreation): add method for myinfo meta in fieldcreation

* refactor(editfielddrawer): refactor fielddrawer to make it more readable

* feat(fieldlistoption): add myinfo field list option

* feat(editmyinfo): add formfielddrawer

* feat(myinfo/base): add shared meta

* feat(myinfo): add remaining fields

* fix(fieldlistoption): update proptypes for field list option

* chore(constants.ts): add myinfo implemented field constants

* fix(editmyinfo): update styling to remove extra divider

* refactor(fieldlistoption): refactor style props into separate method

* feat(myinfo/stories): add stories for myinfo form field

* chore(editfielddrawer): wrap in braces for cleanliness

* chore(mock myinfo constants): shifts constants to msw/mocks

* refactor(fieldcreation): shifts marriagecertno from being number only to shorttextfield

* refactor(fieldcreation): change field options to be injected at runtime

* refactor(editmyinfo): changed to use existing methods

* chore(editmyinfo): update to fit figma design

* chore(myinfo): deletes shared myinfo types and shift them to FE only

* refactor(myinfo): update imports

* fix(editfielddrawer): adds date field back in

* refactor(usebuilderifelds): put augment with myinfo into memo

* fix(constants.ts): fix rebase rewriting myinfo constant

* fix(editmyinfo): omit input transformation so that deps array does not change

* fix(fieldlistoption): fix destructuring css style props

* chore(myinfoattribute): deletes unimplemented types

* chore(myinfo): update usages of impl/unimpl to base myinfoattr

* chore(usebuilderifelds): simplify map

* fix(mocks/form): revert change to mock form field and add new mock with myinfo

* feat(form-v2): add myinfo inline messages (#3948)

* feat(fieldlistdrawer): add component showing myinfo messages

* refactor(fieldlistdrawer): add usememo and conditionally return null

* chore(myinfo link): add href to form guide

timotheeg reviewed (approved):
lgtm

seaerchin committed:
feat(form-v2): MyInfo prefill values (#4028)

* fix(types/ismyinfo): removed types that are unused in myinfo

* feat(augmentwithmyinfo): add myinfo prefilled values

* refactor(templates/field): refactor so myinfoable fields are prefilled

* refactor(ismyinfofieldfield): shard into own file

* Revert "refactor(templates/field): refactor so myinfoable fields are prefilled"

This reverts commit 24c79b80db60e03555376d5e00a22d09ba876a0b.

* chore(publicformpage): add myinfo stories

* revert(augmentwithmyinfo): revert changes

* refactor(myinfo): add prefilled values to public form and admin form

* chore(types/field): add new myinfo type that guarantees prefill

* chore(myinfo/utils): add util method to extract prefilled value into correct format

* refactor(formfields/fieldrowcontainer): refactor to use new util methods to inject prefilled values

* fix(extractprefilledvalue): fixed typo leading to error

* chore(publicformpage.stories): update story to use only myinfo fields

* chore(extractpreviewvalue): changed file name

* chore(myinfo): change prefilledValue -> previewValue

* refactor(hasexistingvalue): shift has existing value into myinfo

* refactor(fieldrow/formfields): update to new api

* chore(mocks/form): update MOCK_MYINFO_FIELDS

* chore(augmentwithmyinfopreview): cjhanged to preview from prefill

* chore(augmentwithmyinfo): rename file for clarity

* fix(newlogicblock.stories): add initial delay so chromatic captures snapshoit properly
